cpython-windows/build.py

```diff
@@ -233,26 +237,34 @@ def find_vs_path(path, msvc_version):
     else:
         raise ValueError(f"unsupported Visual Studio version: {msvc_version}")
 
-    p = subprocess.check_output(
-        [
-            str(vswhere),
-            # Visual Studio 2019.
-            "-version",
-            version,
-            "-property",
-            "installationPath",
-            "-products",
-            "*",
-        ]
-    )
+    cmdline = [
+        str(vswhere),
+        # Visual Studio 2019.
+        "-version",
+        version,
+        "-property",
+        "installationPath",
+        "-prerelease",
+        "-products",
+        "*",
+    ]
+    p = subprocess.check_output(cmdline)
+    if not p:
+        cmd = " ".join('"%s"' % i for i in cmdline)
+        print("could not find visual studio (using '%s')" % cmd)
+        sys.exit(1)
 
-    # Strictly speaking the output may not be UTF-8.
-    p = pathlib.Path(p.strip().decode("utf-8"))
+    p = p.strip().decode("utf-8")
+    if "\n" in p:  # more than one line
+        print("found multiple matching instances of visual studio %s:" % msvc_version)
```

Comment:
So this is like: you have 2022 and the 2022 pre-release installed? Should we just... pick one? (If you know) are they ordered?

Reply:
I was a little surprised to find someone else fighting a similar fight with vs version stuff. The studio I work for has a python-based wrapper (generate.py) around cmake for buffering our non-tech employees from having to learn how to drive cmake for all our various SKUs and environments, and I'd had to build a bunch of stuff into that dealing with vswhere and vcvars.

Unfortunately, about 14 months ago I found a complete workaround: switching to powershell (*) and using the Visual Studio DevShell module's Enter-VsDevShell command.

I never figured out what the difference was between _that_ and just opening a developer cmd prompt was. But with the cmd version, the cmake our wrapper invokes can't find a visual studio and I end up with a bunch of python to find it by hand and inject the variables into the env; using the ps approach just works.

But - I was fighting _all_ the fires at the time, including a Python 2 -> 3 migration, at the start of the work so it's possible, probable?, I just missed something really trivial.
 
(* we were resurrecting the Windows build of 15+ year old code that had spent 8 years mostly on Mac; Open Source + Cross-Platform powershell had just come out; I didn't know squat about Jenkins and I had 30 years experience with unix shells and that-time-I-played-jedi-knight total experience with dos. I figured using PS as a rosetta stone would help me find my feet or give me great material for a blong-rant... It turned out to be like going from Perl to Python. I actively switch between mac/win/lin these days, because I can take my scripts with me without any resistance)

Reply:
For context, my current machine reports this if I remove the `-version` argument:

```
E:\>"C:\Program Files (x86)\Microsoft Visual Studio\Installer\vswhere.exe" -property installationPath -prerelease -products *
C:\Program Files\Microsoft Visual Studio\2022\Preview
C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools
```

I definitely installed Vs2022 preview after BuildTools, but I did just updated Vs2022 preview, and I couldn't find documentation on the ordering.

Reply:
There are two options:
```
  -latest        Return only the newest version and last installed.
  -sort          Sorts the instances from newest version and last installed to oldest.
                 When used with "find", first instances are sorted then files are sorted lexigraphically.
```

outputs:

```
E:\>"C:\Program Files (x86)\Microsoft Visual Studio\Installer\vswhere.exe" -property installationPath -prerelease -products *
C:\Program Files\Microsoft Visual Studio\2022\Preview
C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools

E:\>"C:\Program Files (x86)\Microsoft Visual Studio\Installer\vswhere.exe" -property installationPath -prerelease -products * -sort
C:\Program Files\Microsoft Visual Studio\2022\Preview
C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools

E:\>"C:\Program Files (x86)\Microsoft Visual Studio\Installer\vswhere.exe" -property installationPath -prerelease -products * -latest
C:\Program Files\Microsoft Visual Studio\2022\Preview
```

ergo we could use '-latest', but that's not really solving the problem.

It would be nice if running this from a devshell could just use the _current_ visual studio, but I lost that fight several times trying to do something similar with our studio's python-based cmake wrapper and haven't had the courage to revisit it since finding using powershell was a sufficient workaround =/

Reply:
Sorry, can you clarify why using `-latest` isn't really solving the problem?

Reply:
If I've been building time-and-again with C++X language features only available in the Preview edition and then it suddenly fails on me because it transparently switched to using the latest update of the Build Tools release, that's going to be a head scratcher, especially if it was an automatic update nobody manually triggered.

If they have multiple installs of the same edition, I'm sure they're expecting to have a switch of some kind for it, but I don't really want to explode *this* diff into tackling that.

It's not uncommon for canary builders or ci machines to have multiple VS variants of the same edition installed: 1/ VS 2022 Build Tools, 2/ VS 2022 Professional, 3/ VS 2022 Ultimate Preview, or if they have 4/ a Community Edition installed that gives them licensing considerations.